eth/downloader/downloader_test.go

```diff
@@ -1753,3 +1754,81 @@ func testDeliverHeadersHang(t *testing.T, protocol int, mode SyncMode) {
 		tester.downloader.peers.peers["peer"].peer.(*floodingTestPeer).pend.Wait()
 	}
 }
+
+func TestRemoteHeaderRequestSpan(t *testing.T) {
+
+	testCases := []struct {
+		remoteHeight uint64
+		localHeight  uint64
+		expected     []int
+	}{
+		// Remote is way higher. We should ask for the remote head and go backwards
+		{1500, 1000,
+			[]int{1323, 1339, 1355, 1371, 1387, 1403, 1419, 1435, 1451, 1467, 1483, 1499},
+		},
+		{15000, 13006,
+			[]int{14823, 14839, 14855, 14871, 14887, 14903, 14919, 14935, 14951, 14967, 14983, 14999},
+		},
```

Comment:
Seems weird that in normal mode of operation we would request headers we most certainly don't have yet. Does it really make sense to ask for ancestors way out of our chain height?

Reply:
That's a very good question. In non-normal mode of operation, there's this odd thing that may happen:
```
      common ancestor A (1000)
                    /                     \
    ourchain B(1001)      otherchainC(1001)
       |                                      |
       |                                      | 
    ourhead E(1200)        otherblock G(1200)
                                 otherhead G(1500)
```

So with the current downloader logic, the first time it sees the (bad) sidechain, it will 

* download from `A` to `G`, and try to insert 500 blocks ( or so) . 
* If ancestors are pruned, it will write `C` to `G` to disk, and then fail when it tries to process `G` for `td`-reasons. 

Next time, it will
* mistakenly believe `G` (or a few blocks before) is the common ancestor, since it's present on disk. It will again download `G` to `F`, and fail again. 
It will then repeat this pattern over and over again. 

The change in the downloader is to _not_ view common ancestor as indeed an ancestor, but merely 'the latest block that we already have on disk'. Coupled with the changes in block processor, the new behaviour at the scenario above is. 
 
* download from `A` to `G`, and try to insert 500 blocks ( or so) . 
* If ancestors are pruned, it will write `C` to `F` to disk, and then fail when it tries to process `G` for `td`-reasons. 

Next time, it will
* identify `F` (or some block right before) as the common ancestor, and only download a couple of blocks. 

(It will then repeat this pattern over and over again still)


So how will it behave in common scenarios? 
If we're synching from scratch, it will fail, and have to fall back to a binary search. However, the binary search is actually not that bad, there are more roundtrips but at least we hit the exact right block, instead of missing and downloading ~10 extra blocks.

Reply:
Separating the downloader-change out from the blockchain import change might require changing this behaviour, however, so that the span of blocks always contain our own height.